src/crate/client/result.py

```diff
@@ -0,0 +1,66 @@
+import typing as t
+from functools import cached_property
+
+
+class BulkResultItem(t.TypedDict):
+    """
+    Define the shape of a CrateDB bulk request response item.
+    """
+
+    rowcount: int
+
+
+class BulkResponse:
+    """
+    Manage CrateDB bulk request responses.
+    Accepts a list of bulk arguments (parameter list) and a list of bulk response items.
+
+    https://cratedb.com/docs/crate/reference/en/latest/interfaces/http.html#bulk-operations
+    """
+
+    def __init__(
+            self,
+            records: t.Union[t.Iterable[t.Dict[str, t.Any]], None],
+            results: t.Union[t.Iterable[BulkResultItem], None]):
+        self.records = records
+        self.results = results
+
+    @cached_property
+    def failed_records(self) -> t.List[t.Dict[str, t.Any]]:
+        """
+        Compute list of failed records.
+
+        CrateDB signals failed inserts using `rowcount=-2`.
+
+        https://cratedb.com/docs/crate/reference/en/latest/interfaces/http.html#error-handling
+        """
+        if self.records is None or self.results is None:
```

Comment:
I'd rather raise errors (even separate errors in this case. (meaning if self.records is "None" one error, if self.results is "None", a separate error.

Reply:
The callsite at [^1] likes to have the situation handled gracefully, but maybe that is just because I wanted to have it "convenient". Please poke me another time if you think the component should really raise an exception in this situation, and I will be happy to adjust.

[^1]: https://github.com/crate/cratedb-toolkit/blob/v0.0.27/cratedb_toolkit/io/core.py#L147-L149

Reply:
imho, and if I understand it correctly those 2 shouldn't be none unless something has gone really wrong. So why not handle it properly, catching the exceptions in the toolkit if needed?

Reply:
Improved with 4da893a73720, thanks!